File: leet-code/349. Intersection of Two Arrays/IntersectionOfTwoArrays349_2.java
package com.example;

import java.util.*;

public class IntersectionOfTwoArrays349_2 {
    public int[] intersection(int[] nums1, int[] nums2) {
        Arrays.sort(nums2);
        Set<Integer> answer = new HashSet<>();
        for (int num : nums1) if (containsInt(nums2, num)) answer.add(num);
        return answer.stream().mapToInt(i -> i).toArray();
    }

    private boolean containsInt(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return true;
            if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
}

File: leet-code/349. Intersection of Two Arrays/IntersectionOfTwoArrays349_2.kt
package com.example

class IntersectionOfTwoArrays349_2 {
    fun intersection(nums1: IntArray, nums2: IntArray): IntArray {
        nums1.sort()
        val set = mutableSetOf<Int>()
        nums2.forEach { if (exist(nums1, it)) set.add(it) }
        return set.toIntArray()
    }

    private fun exist(arr: IntArray, num: Int): Boolean {
        var low = 0
        var high = arr.lastIndex
        while (low <= high) {
            val mid = low + (high - low) / 2
            when {
                arr[mid] < num -> low = mid + 1
                arr[mid] > num -> high = mid - 1
                else -> return true
            }
        }
        return false
    }
}
